Lib/_colorize.py

```diff
@@ -138,26 +286,37 @@ def can_colorize(*, file: IO[str] | IO[bytes] | None = None) -> bool:
         return hasattr(file, "isatty") and file.isatty()
 
 
-def set_theme(t: dict[ColorTag, str] | None = None) -> None:
-    global theme
+default_theme = Theme()
+theme_no_color = default_theme.no_colors()
+
+
+def get_theme(
+    *,
+    tty_file: IO[str] | IO[bytes] | None = None,
+    force_color: bool = False,
+    force_no_color: bool = False,
+) -> Theme:
+    """Returns the currently set theme, potentially in a zero-color variant.
+
+    In cases where colorizing is not possible (see `can_colorize`), the returned
+    theme contains all empty strings in all color definitions.
+    See `Theme.no_colors()` for more information.
+
+    It is recommended not to cache the result of this function for extended
+    periods of time because the user might influence theme selection by
+    the interactive shell, a debugger, or application-specific code. The
+    environment (including environment variable state and console configuration
+    on Windows) can also change in the course of the application life cycle.
+    """
+    if force_color or (not force_no_color and can_colorize(file=tty_file)):
```

Comment:
This thing is being called in loops and potentially hot paths and is constantly checking `can_colorize` which is suboptimal as is checking environment variables, checking if is a tty and whatnot. Any reason we don't cache this function?

Reply:
Explained in the docstring:

> It is recommended not to cache the result of this function for extended periods of time because the user might influence theme selection by the interactive shell, a debugger, or application-specific code. The environment (including environment variable state and console configuration on Windows) can also change in the course of the application life cycle.

The third-party `termcolor` package (63m downloads/month) has very similar function to `can_colorize()` and we just wrapped it in `@cache`: https://github.com/termcolor/termcolor/pull/107. Only released five days ago, no negative feedback so far.

Reply:
> If you cached this in the REPL, you wouldn't be able to interactively see changes on coloring based on changes to `os.environ` or Windows console control commands.

I don't know if I think this is a good reason not to cache it by default

Reply:
You could add a 'force_refresh' or something